File: cryptoheir-rs/src/commands/prepare.rs
//! Prepare command - creates unsigned transactions (requires network access)

use crate::{contract, network, qr, types::*, Result};
use alloy::primitives::{Address, U256};
use clap::Subcommand;
use tracing::info;

// Type alias for the RPC client type
type RpcClient = alloy::providers::RootProvider<alloy::transports::http::Http<alloy::transports::http::Client>>;

#[derive(Subcommand, Debug, Clone)]
pub enum Operation {
    /// Deploy a new CryptoHeir contract
    Deploy,

    /// Deposit funds into an inheritance
    Deposit {
        /// Beneficiary address
        #[arg(short, long)]
        beneficiary: Address,

        /// Amount to deposit (in ETH or token units)
        #[arg(short, long)]
        amount: String,

        /// Deadline (Unix timestamp in seconds)
        #[arg(short, long)]
        deadline: u64,

        /// ERC20 token address (if depositing tokens instead of ETH)
        #[arg(short, long)]
        token: Option<Address>,

        /// Contract address
        #[arg(short, long, env = "CONTRACT_ADDRESS")]
        contract: Option<Address>,
    },

    /// Claim an inheritance (as beneficiary)
    Claim {
        /// Inheritance ID
        #[arg(short, long)]
        id: U256,

        /// Contract address
        #[arg(short, long, env = "CONTRACT_ADDRESS")]
        contract: Option<Address>,
    },

    /// Reclaim an inheritance (as owner, before deadline)
    Reclaim {
        /// Inheritance ID
        #[arg(short, long)]
        id: U256,

        /// Contract address
        #[arg(short, long, env = "CONTRACT_ADDRESS")]
        contract: Option<Address>,
    },

    /// Extend the deadline for an inheritance
    ExtendDeadline {
        /// Inheritance ID
        #[arg(short, long)]
        id: U256,

        /// New deadline (Unix timestamp in seconds)
        #[arg(short, long)]
        new_deadline: u64,

        /// Contract address
        #[arg(short, long, env = "CONTRACT_ADDRESS")]
        contract: Option<Address>,
    },

    /// Transfer fee collector role (initiate 2-step transfer)
    TransferFeeCollector {
        /// New fee collector address
        #[arg(short, long)]
        new_collector: Address,

        /// Contract address
        #[arg(short, long, env = "CONTRACT_ADDRESS")]
        contract: Option<Address>,
    },

    /// Accept fee collector role (complete 2-step transfer)
    AcceptFeeCollector {
        /// Contract address
        #[arg(short, long, env = "CONTRACT_ADDRESS")]
        contract: Option<Address>,
    },
}

pub async fn execute(
    operation: Operation,
    network: Option<String>,
    rpc_url: Option<String>,
    output: String,
    generate_qr: bool,
) -> Result<()> {
    info!("Preparing transaction...");

    // Load configuration
    let config = Config::load()?;

    // Determine RPC URL
    let rpc_url = rpc_url
        .or(config.rpc_url)
        .or_else(|| {
            network::get_rpc_url(
                network.as_deref().unwrap_or("sepolia"),
                config.infura_api_key.as_deref(),
            )
        })
        .ok_or_else(|| eyre::eyre!("No RPC URL provided or configured"))?;

    info!("Connecting to network via {}", rpc_url);

    // Get signer address
    let signer_address = config
        .signer_address
        .ok_or_else(|| eyre::eyre!("SIGNER_ADDRESS not set in environment"))?;

    // Create RPC client
    let client = network::create_client(&rpc_url).await?;

    // Get network info
    let chain_id = network::get_chain_id(&client).await?;
    let network_name = network.unwrap_or_else(|| "custom".to_string());

    info!(
        "Connected to {} (chain ID: {})",
        network_name, chain_id
    );

    // Get nonce
    let nonce = network::get_nonce(&client, signer_address).await?;
    info!("Nonce: {}", nonce);

    // Prepare transaction based on operation
    let tx_params = match operation {
        Operation::Deploy => {
            prepare_deploy(&client, signer_address, nonce, chain_id, &network_name).await?
        }
        Operation::Deposit {
            beneficiary,
            amount,
            deadline,
            token,
            contract,
        } => {
            let contract_addr = contract.or(config.contract_address).ok_or_else(|| {
                eyre::eyre!("Contract address required (use --contract or set CONTRACT_ADDRESS)")
            })?;
            prepare_deposit(
                &client,
                signer_address,
                nonce,
                chain_id,
                &network_name,
                contract_addr,
                beneficiary,
                amount,
                deadline,
                token,
            )
            .await?
        }
        Operation::Claim { id, contract } => {
            let contract_addr = contract.or(config.contract_address).ok_or_else(|| {
                eyre::eyre!("Contract address required (use --contract or set CONTRACT_ADDRESS)")
            })?;
            prepare_claim(
                &client,
                signer_address,
                nonce,
                chain_id,
                &network_name,
                contract_addr,
                id,
            )
            .await?
        }
        Operation::Reclaim { id, contract } => {
            let contract_addr = contract.or(config.contract_address).ok_or_else(|| {
                eyre::eyre!("Contract address required (use --contract or set CONTRACT_ADDRESS)")
            })?;
            prepare_reclaim(
                &client,
                signer_address,
                nonce,
                chain_id,
                &network_name,
                contract_addr,
                id,
            )
            .await?
        }
        Operation::ExtendDeadline {
            id,
            new_deadline,
            contract,
        } => {
            let contract_addr = contract.or(config.contract_address).ok_or_else(|| {
                eyre::eyre!("Contract address required (use --contract or set CONTRACT_ADDRESS)")
            })?;
            prepare_extend_deadline(
                &client,
                signer_address,
                nonce,
                chain_id,
                &network_name,
                contract_addr,
                id,
                new_deadline,
            )
            .await?
        }
        Operation::TransferFeeCollector {
            new_collector,
            contract,
        } => {
            let contract_addr = contract.or(config.contract_address).ok_or_else(|| {
                eyre::eyre!("Contract address required (use --contract or set CONTRACT_ADDRESS)")
            })?;
            prepare_transfer_fee_collector(
                &client,
                signer_address,
                nonce,
                chain_id,
                &network_name,
                contract_addr,
                new_collector,
            )
            .await?
        }
        Operation::AcceptFeeCollector { contract } => {
            let contract_addr = contract.or(config.contract_address).ok_or_else(|| {
                eyre::eyre!("Contract address required (use --contract or set CONTRACT_ADDRESS)")
            })?;
            prepare_accept_fee_collector(
                &client,
                signer_address,
                nonce,
                chain_id,
                &network_name,
                contract_addr,
            )
            .await?
        }
    };

    // Save to file
    let json = serde_json::to_string_pretty(&tx_params)?;
    std::fs::write(&output, &json)?;

    info!("Transaction parameters saved to {}", output);
    println!("\n✓ Transaction prepared successfully!");
    println!("  Output: {}", output);
    println!("  Network: {} (chain ID: {})", network_name, chain_id);
    println!("  Estimated cost: {} ETH", tx_params.metadata.estimated_cost);

    // Generate QR code if requested
    if generate_qr {
        info!("Generating QR code...");
        qr::display_qr(&json)?;
    }

    println!("\nNext step: Transfer {} to offline machine and run:", output);
    println!("  cryptoheir-rs sign -i {}", output);

    Ok(())
}

async fn prepare_deploy(
    client: &RpcClient,
    from: Address,
    nonce: u64,
    chain_id: u64,
    network_name: &str,
) -> Result<TxParams> {
    info!("Preparing contract deployment...");

    // Load contract bytecode
    let bytecode = contract::load_bytecode()?;
    let data = bytecode.into();

    // Estimate gas
    let gas_limit = network::estimate_gas(client, from, None, &data, None).await?;

    // Get gas prices
    let (max_fee_per_gas, max_priority_fee_per_gas, gas_price) =
        network::get_gas_prices(client).await?;

    // Determine transaction type and set appropriate gas fields
    let (tx_type, final_max_fee, final_priority_fee, final_gas_price) =
        if max_fee_per_gas.is_some() {
            // EIP-1559 transaction (type 2)
            (2, max_fee_per_gas, max_priority_fee_per_gas, None)
        } else {
            // Legacy transaction (type 0)
            (0, None, None, gas_price)
        };

    // Calculate estimated cost
    let estimated_cost = if let Some(max_fee) = final_max_fee {
        network::format_eth(gas_limit * max_fee)
    } else if let Some(price) = final_gas_price {
        network::format_eth(gas_limit * price)
    } else {
        "unknown".to_string()
    };

    Ok(TxParams {
        mode: TransactionMode::Deploy,
        function_name: None,
        params: None,
        transaction: TransactionData {
            tx_type,
            from,
            to: None,
            data,
            nonce,
            chain_id,
            gas_limit,
            max_fee_per_gas: final_max_fee,
            max_priority_fee_per_gas: final_priority_fee,
            gas_price: final_gas_price,
            value: None,
        },
        metadata: Metadata {
            network: NetworkInfo {
                name: network_name.to_string(),
                chain_id,
                rpc_url: None,
            },
            estimated_cost,
            timestamp: chrono::Utc::now().to_rfc3339(),
            prepared: true,
            signed: false,
            signed_at: None,
        },
    })
}

async fn prepare_deposit(
    client: &RpcClient,
    from: Address,
    nonce: u64,
    chain_id: u64,
    network_name: &str,
    contract: Address,
    beneficiary: Address,
    amount: String,
    deadline: u64,
    token: Option<Address>,
) -> Result<TxParams> {
    info!("Preparing deposit transaction...");

    // Parse amount
    let amount_wei = alloy::primitives::utils::parse_ether(&amount)?;

    // Encode function call
    let (data, value) =
        contract::encode_deposit(beneficiary, amount_wei, deadline, token).await?;

    // Estimate gas
    let gas_limit = network::estimate_gas(client, from, Some(contract), &data, value).await?;

    // Get gas prices
    let (max_fee_per_gas, max_priority_fee_per_gas, gas_price) =
        network::get_gas_prices(client).await?;

    // Determine transaction type and set appropriate gas fields
    let (tx_type, final_max_fee, final_priority_fee, final_gas_price) =
        if max_fee_per_gas.is_some() {
            // EIP-1559 transaction (type 2)
            (2, max_fee_per_gas, max_priority_fee_per_gas, None)
        } else {
            // Legacy transaction (type 0)
            (0, None, None, gas_price)
        };

    // Calculate estimated cost (including value being sent)
    let gas_cost = if let Some(max_fee) = final_max_fee {
        gas_limit * max_fee
    } else if let Some(price) = final_gas_price {
        gas_limit * price
    } else {
        U256::ZERO
    };
    let total_cost = gas_cost + value.unwrap_or(U256::ZERO);
    let estimated_cost = network::format_eth(total_cost);

    Ok(TxParams {
        mode: TransactionMode::Call,
        function_name: Some("deposit".to_string()),
        params: Some(serde_json::json!({
            "beneficiary": beneficiary,
            "amount": amount,
            "deadline": deadline,
            "token": token,
        })),
        transaction: TransactionData {
            tx_type,
            from,
            to: Some(contract),
            data,
            nonce,
            chain_id,
            gas_limit,
            max_fee_per_gas: final_max_fee,
            max_priority_fee_per_gas: final_priority_fee,
            gas_price: final_gas_price,
            value,
        },
        metadata: Metadata {
            network: NetworkInfo {
                name: network_name.to_string(),
                chain_id,
                rpc_url: None,
            },
            estimated_cost,
            timestamp: chrono::Utc::now().to_rfc3339(),
            prepared: true,
            signed: false,
            signed_at: None,
        },
    })
}

// Stub implementations for other operations
async fn prepare_claim(
    _client: &RpcClient,
    _from: Address,
    _nonce: u64,
    _chain_id: u64,
    _network_name: &str,
    _contract: Address,
    _id: U256,
) -> Result<TxParams> {
    info!("Preparing claim transaction...");
    // TODO: Implement claim encoding
    todo!("prepare_claim not yet implemented")
}

async fn prepare_reclaim(
    _client: &RpcClient,
    _from: Address,
    _nonce: u64,
    _chain_id: u64,
    _network_name: &str,
    _contract: Address,
    _id: U256,
) -> Result<TxParams> {
    info!("Preparing reclaim transaction...");
    // TODO: Implement reclaim encoding
    todo!("prepare_reclaim not yet implemented")
}

async fn prepare_extend_deadline(
    _client: &RpcClient,
    _from: Address,
    _nonce: u64,
    _chain_id: u64,
    _network_name: &str,
    _contract: Address,
    _id: U256,
    _new_deadline: u64,
) -> Result<TxParams> {
    info!("Preparing extend deadline transaction...");
    // TODO: Implement extend_deadline encoding
    todo!("prepare_extend_deadline not yet implemented")
}

async fn prepare_transfer_fee_collector(
    _client: &RpcClient,
    _from: Address,
    _nonce: u64,
    _chain_id: u64,
    _network_name: &str,
    _contract: Address,
    _new_collector: Address,
) -> Result<TxParams> {
    info!("Preparing transfer fee collector transaction...");
    // TODO: Implement transfer_fee_collector encoding
    todo!("prepare_transfer_fee_collector not yet implemented")
}

async fn prepare_accept_fee_collector(
    _client: &RpcClient,
    _from: Address,
    _nonce: u64,
    _chain_id: u64,
    _network_name: &str,
    _contract: Address,
) -> Result<TxParams> {
    info!("Preparing accept fee collector transaction...");
    // TODO: Implement accept_fee_collector encoding
    todo!("prepare_accept_fee_collector not yet implemented")
}

File: cryptoheir-go/internal/commands/prepare.go
package commands

import (
	"context"
	"encoding/json"
	"fmt"
	"math/big"
	"os"
	"time"

	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/contract"
	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/network"
	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/types"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/ethclient"
	"github.com/ethereum/go-ethereum/params"
	"github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
)

var log = logrus.New()

// PrepareCmd represents the prepare command
var PrepareCmd = &cobra.Command{
	Use:   "prepare [deploy|deposit]",
	Short: "Prepare an unsigned transaction for offline signing",
	Long: `Prepare an unsigned transaction by connecting to the network,
estimating gas, and creating a transaction parameters file for offline signing.

Supports:
  - deploy: Deploy a new CryptoHeir contract
  - deposit: Create an inheritance deposit`,
	Args: cobra.ExactArgs(1),
	RunE: runPrepare,
}

var (
	// Common flags
	networkFlag string
	rpcURLFlag  string
	outputFlag  string

	// Deposit flags
	beneficiaryFlag string
	amountFlag      string
	deadlineFlag    int64
	tokenFlag       string
)

func init() {
	// Common flags
	PrepareCmd.PersistentFlags().StringVar(&networkFlag, "network", "sepolia", "Network name (sepolia, mainnet, etc.)")
	PrepareCmd.PersistentFlags().StringVar(&rpcURLFlag, "rpc-url", "", "Custom RPC URL (overrides network)")
	PrepareCmd.PersistentFlags().StringVarP(&outputFlag, "output", "o", "tx-params.json", "Output file path")

	// Deposit-specific flags
	PrepareCmd.PersistentFlags().StringVar(&beneficiaryFlag, "beneficiary", "", "Beneficiary address")
	PrepareCmd.PersistentFlags().StringVar(&amountFlag, "amount", "", "Amount in ETH (e.g., 1.5)")
	PrepareCmd.PersistentFlags().Int64Var(&deadlineFlag, "deadline", 0, "Deadline as Unix timestamp")
	PrepareCmd.PersistentFlags().StringVar(&tokenFlag, "token", "", "ERC20 token address (omit for native ETH)")
}

func runPrepare(cmd *cobra.Command, args []string) error {
	operation := args[0]

	// Load configuration
	config, err := types.LoadConfig()
	if err != nil {
		return fmt.Errorf("failed to load config: %w", err)
	}

	// Initialize contract module
	if err := contract.Initialize(); err != nil {
		return fmt.Errorf("failed to initialize contract: %w", err)
	}

	// Determine RPC URL
	rpcURL := rpcURLFlag
	if rpcURL == "" {
		rpcURL, err = network.GetRPCURL(networkFlag, config.InfuraAPIKey)
		if err != nil {
			return fmt.Errorf("failed to get RPC URL: %w", err)
		}
	}

	// Connect to network
	ctx := context.Background()
	client, err := network.CreateClient(ctx, rpcURL)
	if err != nil {
		return err
	}
	defer client.Close()

	log.Info("Connected to network")

	// Get chain ID
	chainID, err := network.GetChainID(ctx, client)
	if err != nil {
		return err
	}
	log.Infof("Chain ID: %d", chainID)

	// Get signer address
	if config.SignerAddress == nil {
		return fmt.Errorf("SIGNER_ADDRESS not set in environment")
	}
	signerAddress := *config.SignerAddress
	log.Infof("Signer address: %s", signerAddress.Hex())

	// Get nonce
	nonce, err := network.GetNonce(ctx, client, signerAddress)
	if err != nil {
		return err
	}
	log.Infof("Nonce: %d", nonce)

	// Prepare transaction based on operation
	var txParams *types.TxParams
	switch operation {
	case "deploy":
		txParams, err = prepareDeploy(ctx, client, signerAddress, nonce, chainID, networkFlag, rpcURL)
	case "deposit":
		txParams, err = prepareDeposit(ctx, client, config, signerAddress, nonce, chainID, networkFlag, rpcURL)
	default:
		return fmt.Errorf("unsupported operation: %s (supported: deploy, deposit)", operation)
	}

	if err != nil {
		return err
	}

	// Save to file
	data, err := json.MarshalIndent(txParams, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to serialize transaction: %w", err)
	}

	if err := os.WriteFile(outputFlag, data, 0644); err != nil {
		return fmt.Errorf("failed to write output file: %w", err)
	}

	log.Infof("✓ Transaction prepared successfully")
	log.Infof("  Output: %s", outputFlag)
	log.Infof("  Next: Transfer to offline machine and run 'cryptoheir sign -i %s'", outputFlag)

	return nil
}

func prepareDeploy(ctx context.Context, client *ethclient.Client, signerAddress common.Address, nonce uint64, chainID uint64, networkName, rpcURL string) (*types.TxParams, error) {
	log.Info("Preparing contract deployment...")

	// Load bytecode
	bytecode, err := contract.LoadBytecode()
	if err != nil {
		return nil, err
	}
	log.Infof("Contract bytecode loaded (%d bytes)", len(bytecode))

	// Estimate gas
	gasLimit, err := network.EstimateGas(ctx, client, signerAddress, nil, bytecode, nil)
	if err != nil {
		return nil, fmt.Errorf("gas estimation failed: %w", err)
	}
	log.Infof("Estimated gas: %s", gasLimit.String())

	// Get gas prices
	gasPrices, err := network.GetGasPrices(ctx, client)
	if err != nil {
		return nil, err
	}

	// Build transaction data
	txData := types.TransactionData{
		From:     signerAddress,
		To:       nil, // Contract deployment
		Data:     bytecode,
		Nonce:    nonce,
		ChainID:  chainID,
		GasLimit: types.NewBigInt(gasLimit),
	}

	// Set gas prices based on transaction type
	if gasPrices.IsEIP1559 {
		txData.TxType = 2
		txData.MaxFeePerGas = types.NewBigInt(gasPrices.MaxFeePerGas)
		txData.MaxPriorityFeePerGas = types.NewBigInt(gasPrices.MaxPriorityFeePerGas)
		log.Infof("EIP-1559: Max fee %s gwei, Priority %s gwei",
			weiToGwei(gasPrices.MaxFeePerGas), weiToGwei(gasPrices.MaxPriorityFeePerGas))
	} else {
		txData.TxType = 0
		txData.GasPrice = types.NewBigInt(gasPrices.GasPrice)
		log.Infof("Legacy: Gas price %s gwei", weiToGwei(gasPrices.GasPrice))
	}

	// Build metadata
	metadata := types.Metadata{
		PreparedAt:  time.Now().UTC().Format(time.RFC3339),
		Network:     types.NetworkInfo{Name: networkName, ChainID: chainID, RPCURL: rpcURL},
		ToolVersion: "cryptoheir-go v0.1.0",
	}

	// Build TxParams
	txParams := &types.TxParams{
		Mode:        types.TransactionModeDeploy,
		Transaction: txData,
		Metadata:    metadata,
	}

	return txParams, nil
}

func prepareDeposit(ctx context.Context, client *ethclient.Client, config *types.Config, signerAddress common.Address, nonce uint64, chainID uint64, networkName, rpcURL string) (*types.TxParams, error) {
	log.Info("Preparing deposit transaction...")

	// Validate required flags
	if beneficiaryFlag == "" {
		return nil, fmt.Errorf("--beneficiary is required")
	}
	if amountFlag == "" {
		return nil, fmt.Errorf("--amount is required")
	}
	if deadlineFlag == 0 {
		return nil, fmt.Errorf("--deadline is required")
	}

	// Parse beneficiary address
	beneficiary := common.HexToAddress(beneficiaryFlag)
	if beneficiary == (common.Address{}) {
		return nil, fmt.Errorf("invalid beneficiary address")
	}

	// Parse amount (ETH to wei)
	amount, err := parseEther(amountFlag)
	if err != nil {
		return nil, fmt.Errorf("invalid amount: %w", err)
	}

	// Parse deadline
	deadline := big.NewInt(deadlineFlag)

	// Parse token address (optional)
	var token *common.Address
	if tokenFlag != "" {
		tokenAddr := common.HexToAddress(tokenFlag)
		token = &tokenAddr
	}

	// Get contract address
	if config.ContractAddress == nil {
		return nil, fmt.Errorf("CONTRACT_ADDRESS not set in environment")
	}
	contractAddress := *config.ContractAddress

	// Encode deposit function
	data, value, err := contract.EncodeDeposit(beneficiary, amount, deadline, token)
	if err != nil {
		return nil, err
	}

	// Estimate gas
	gasLimit, err := network.EstimateGas(ctx, client, signerAddress, &contractAddress, data, value)
	if err != nil {
		return nil, fmt.Errorf("gas estimation failed: %w", err)
	}
	log.Infof("Estimated gas: %s", gasLimit.String())

	// Get gas prices
	gasPrices, err := network.GetGasPrices(ctx, client)
	if err != nil {
		return nil, err
	}

	// Build transaction data
	txData := types.TransactionData{
		From:     signerAddress,
		To:       &contractAddress,
		Data:     data,
		Nonce:    nonce,
		ChainID:  chainID,
		GasLimit: types.NewBigInt(gasLimit),
	}

	// Set value if native token
	if value != nil {
		txData.Value = types.NewBigInt(value)
		log.Infof("Deposit value: %s", network.FormatEth(value))
	}

	// Set gas prices
	if gasPrices.IsEIP1559 {
		txData.TxType = 2
		txData.MaxFeePerGas = types.NewBigInt(gasPrices.MaxFeePerGas)
		txData.MaxPriorityFeePerGas = types.NewBigInt(gasPrices.MaxPriorityFeePerGas)
	} else {
		txData.TxType = 0
		txData.GasPrice = types.NewBigInt(gasPrices.GasPrice)
	}

	// Build parameters JSON
	params := map[string]interface{}{
		"beneficiary": beneficiary.Hex(),
		"amount":      amount.String(),
		"deadline":    deadline.String(),
	}
	if token != nil {
		params["token"] = token.Hex()
	}
	paramsJSON, _ := json.Marshal(params)

	// Build metadata
	metadata := types.Metadata{
		PreparedAt:  time.Now().UTC().Format(time.RFC3339),
		Network:     types.NetworkInfo{Name: networkName, ChainID: chainID, RPCURL: rpcURL},
		ToolVersion: "cryptoheir-go v0.1.0",
	}

	// Build TxParams
	txParams := &types.TxParams{
		Mode:         types.TransactionModeCall,
		FunctionName: "deposit",
		Params:       paramsJSON,
		Transaction:  txData,
		Metadata:     metadata,
	}

	log.Infof("Deposit prepared for beneficiary %s", beneficiary.Hex())
	return txParams, nil
}

// parseEther converts an ETH string to wei (*big.Int)
func parseEther(ethStr string) (*big.Int, error) {
	// Parse as float first
	var ethFloat float64
	if _, err := fmt.Sscanf(ethStr, "%f", &ethFloat); err != nil {
		return nil, fmt.Errorf("invalid number format: %w", err)
	}

	// Convert to wei (multiply by 1e18)
	weiFloat := ethFloat * params.Ether
	wei := new(big.Int)
	wei.SetUint64(uint64(weiFloat))

	return wei, nil
}

// weiToGwei converts wei to gwei string
func weiToGwei(wei *big.Int) string {
	if wei == nil {
		return "0"
	}
	gwei := new(big.Float).Quo(new(big.Float).SetInt(wei), big.NewFloat(params.GWei))
	return fmt.Sprintf("%.2f", gwei)
}
